src/rust/engine/process_execution/sandboxer/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+        let inactivity_counter = Arc::new(AtomicUsize::new(0));
+        let inactivity_counter_ref = inactivity_counter.clone();
+        if with_shutdown_thread {
+            thread::spawn(move || {
+                loop {
+                    // Note: just a regular thread::sleep() since this is not an async context.
+                    thread::sleep(SandboxerService::POLLING_INTERVAL);
+                    if !fs::exists(&socket_path).unwrap_or(false) {
+                        warn!("Socket file {} deleted. Exiting.", &socket_path.display());
+                        process::exit(22);
+                    }
+                    if inactivity_counter_ref.fetch_add(1, Ordering::Relaxed)
+                        > Self::ALLOWED_IDLE_BEATS
+                    {
+                        warn!("Idle time limit exceeded. Exiting.");
+                        process::exit(23);
+                    }
+                }
+            });
+        }
+
+        info!(
+            "Sandboxer server started on socket path {}",
+            self.socket_path.display()
+        );
+        let store = self
+            .store_cli_opt
+            .create_store(task_executor::Executor::new())
+            .await?;
+        let router = Server::builder().add_service(SandboxerGrpcServer::new(SandboxerGrpcImpl {
+            store,
+            inactivity_counter,
+        }));
+        router
+            .serve_with_incoming(UnixListenerStream::new(listener))
+            .await?;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct SandboxerGrpcImpl {
+    store: Store,
+
+    // We increment this at every poll interval, and reset it on every gRPC call.
+    inactivity_counter: Arc<AtomicUsize>,
+}
+
+impl SandboxerGrpcImpl {
+    async fn do_materialize_directory(
+        &self,
+        request: Request<MaterializeDirectoryRequest>,
+    ) -> Result<Response<MaterializeDirectoryResponse>, String> {
+        let r = request.into_inner();
+        debug!("Received materialize_directory() request: {:#?}", &r);
+        let destination_root: PathBuf = r.destination_root.into();
+        let digest: Digest = r.digest.ok_or("No digest provided")?.try_into()?;
+        let dir_digest = DirectoryDigest::from_persisted_digest(digest);
+        let mutable_paths = BTreeSet::<RelativePath>::from_iter(
+            r.mutable_paths
+                .iter()
+                .map(RelativePath::new)
+                .collect::<Result<Vec<_>, _>>()?
+                .into_iter(),
+        );
+        self.store
+            .materialize_directory(
+                r.destination.into(),
+                &destination_root,
+                dir_digest,
+                false,
+                &mutable_paths,
+                Permissions::Writable,
+            )
+            .await
+            .map_err(|e| e.to_string())?;
+        Ok(Response::new(MaterializeDirectoryResponse {
+            confirmation: { "OK".to_string() },
```

Comment:
Is an OK response like this really necessary? Just returning a normal gRPC status OK would be sufficient, right? (I.e., it would be an empty response message for now.)

Reply:
Not really necessary, we could return nothing in the response.